src/dmd/optimize.d

```diff
@@ -558,6 +558,41 @@ Expression Expression_optimize(Expression e, int result, bool keepLvalue)
         }
         else if (auto ae = e.e1.isIndexExp())
         {
+            if (ae.e2.isIntegerExp() && ae.e1.isIndexExp())
+            {
+                /* Rewrite `(a[i])[index]` to `(&a[i]) + index*size`
```

Comment:
The comment should say why this is done

Reply:
The reason is the same as for every other rewrite - to put it in simpler terms that the rest of the code can handle. I.e. "lowering".